File: java/src/main/java/io/fliqa/example/webhook/WebHookUtils.java
package io.fliqa.example.webhook;

import javax.crypto.*;
import javax.crypto.spec.*;
import java.math.*;
import java.security.*;
import java.util.*;

/**
 * Signature is created by using SHA256 to create a hash
 * Input is: {time-in-epoch-seconds}.{hookUrl}.{recieved_body}
 * Header holds: {time-in-epoch-seconds} and created hash as: t={time-in-epoch-seconds},v={hashHexEncoded}
 * alternately header can hold old and new verification if secret was changed within 24h
 * as: t={time-in-epoch-seconds},v={hashHexEncoded},v0={oldHashHexEncoded}
 */
public class WebHookUtils {

    private static final String DIGEST = "HmacSHA256";
    private static final String TIME_STAMP = "t=";
    private static final String VERIFICATION = "v=";
    private static final String OLD_VERIFICATION = "v0=";

    public static boolean checkSignature(String signature, String secret, String oldSecret, String hookUrl, String body) {

        // Split signature to t={time},v={verification},v0={old_verification}
        String[] timeAndSignature = signature.split(",");
        if (timeAndSignature.length < 2 || timeAndSignature.length > 3) {
            throw new IllegalArgumentException(String.format("Invalid signature, expected time and verification but got: '%s'!", signature));
        }

        Long time = getSignatureTime(timeAndSignature);

        String oldVerification = null;
        String oldCompare = null;
        if (timeAndSignature.length == 3) { // there is the old signature present (double check)
            oldVerification = getOldSignatureVerification(timeAndSignature);
            oldCompare = sign(oldSecret, time.toString(), hookUrl, body);
        }

        String verification = getSignatureVerification(timeAndSignature);
        String compare = sign(secret, time.toString(), hookUrl, body);

        return verification.equals(compare) || (oldVerification != null && oldVerification.equals(oldCompare));
    }

    protected static String sign(String secret, String time, String hookUrl, String content) {
        String input = String.format("%s.%s.%s", time, hookUrl, content);

        try {
            Mac mac = Mac.getInstance(DIGEST);
            mac.init(new SecretKeySpec(secret.getBytes(), DIGEST));
            return toHexString(mac.doFinal(input.getBytes()));
        } catch (InvalidKeyException | NoSuchAlgorithmException e) {
            // Failed to create SHA signature!
            return "";
        }
    }

    public static String toHexString(byte[] arg) {
        return String.format("%x", new BigInteger(1, arg));
    }

    protected static Long getSignatureTime(String[] timeAndSignature) {
        Optional<String> found = Arrays.stream(timeAndSignature).filter(item -> item.toLowerCase().startsWith(TIME_STAMP)).findFirst();
        if (found.isPresent()) {
            try {
                return Long.parseLong(found.get().substring(TIME_STAMP.length()));
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException(String.format("Signature time (%s<timestamp>) invalid: '%s'!", TIME_STAMP, found.get()));
            }
        }

        throw new IllegalArgumentException(String.format("Signature time (%s<timestamp>) missing!", TIME_STAMP));
    }

    protected static String getSignatureVerification(String[] timeAndSignature) {
        Optional<String> found = Arrays.stream(timeAndSignature).filter(item -> item.toLowerCase().startsWith(VERIFICATION)).findFirst();
        return found.map(s -> s.substring(VERIFICATION.length())).orElseThrow(() -> new IllegalArgumentException(String.format("Signature verification (%s<verification>) missing!", VERIFICATION)));
    }

    protected static String getOldSignatureVerification(String[] timeAndSignature) {
        Optional<String> found = Arrays.stream(timeAndSignature).filter(item -> item.toLowerCase().startsWith(OLD_VERIFICATION)).findFirst();
        return found.map(s -> s.substring(OLD_VERIFICATION.length())).orElseThrow(() -> new IllegalArgumentException(String.format("Signature old verification (%s<verification>) missing!", OLD_VERIFICATION)));
    }
}

File: php/webhook-utils.php
<?php

class WebHookUtils
{
    private const DIGEST = 'sha256';

    public static function checkSignature(string $signature, string $secret, string $oldSecret, string $hookUrl, string $body): bool
    {
        // Split signature to t={time},v={verification},v0={old_verification}
        $parts = explode(',', $signature);
        if (count($parts) < 2 || count($parts) > 3) {
            throw new InvalidArgumentException(sprintf(
                "Invalid signature, expected time and verification but got: '%s'!",
                $signature
            ));
        }

        $time = self::getSignatureTime($parts);

        $oldVerification = null;
        $oldCompare = null;
        if (count($parts) === 3) {
            $oldVerification = self::getOldSignatureVerification($parts);
            $oldCompare = self::sign($oldSecret, $time, $hookUrl, $body);
        }

        $verification = self::getSignatureVerification($parts);
        $compare = self::sign($secret, $time, $hookUrl, $body);

        // It might be that we have not updated the secret jet, but on the other side it was already updated
        return $verification === $compare || $oldVerification === $compare || $verification === $oldCompare || $oldVerification === $oldCompare;
    }

    protected static function sign(string $secret, string $time, string $hookUrl, string $content): string
    {
        $input = sprintf('%s.%s.%s', $time, $hookUrl, $content);
        return hash_hmac(self::DIGEST, $input, $secret);
    }

    protected static function getSignatureTime(array $parts): string
    {
        foreach ($parts as $part) {
            if (str_starts_with($part, 't='))
                return substr($part, 2);
        }

        throw new InvalidArgumentException('Missing time (t=) in signature.');
    }

    protected static function getSignatureVerification(array $parts): string
    {
        foreach ($parts as $part) {
            if (str_starts_with($part, 'v='))
                return substr($part, 2);
        }

        throw new InvalidArgumentException('Missing verification (v=) in signature.');
    }

    protected static function getOldSignatureVerification(array $parts): string
    {
        foreach ($parts as $part) {
            if (str_starts_with($part, 'v0='))
                return substr($part, 3);
        }

        throw new InvalidArgumentException('Missing old verification (v0=) in signature.');
    }
}
